File: Pathfinding Visualizer/algorithms/dfs.js
function runDFS() {
  let stack = [start];
  let visited = new Set();
  let parent = {};
  visited.add(start.toString());

  while (stack.length > 0) {
    let [r,c] = stack.pop();
    if (r===end[0] && c===end[1]) {
      reconstructPath(parent);
      return;
    }
    for (let [dr,dc] of [[1,0],[-1,0],[0,1],[0,-1]]) {
      let nr=r+dr, nc=c+dc;
      if (nr>=0 && nr<rows && nc>=0 && nc<cols &&
          !grid[nr][nc].classList.contains("wall") &&
          !visited.has([nr,nc].toString())) {
        stack.push([nr,nc]);
        visited.add([nr,nc].toString());
        parent[[nr,nc]] = [r,c];
        grid[nr][nc].classList.add("visited");
      }
    }
  }
}

File: Pathfinding Visualizer/PathfindingVisualizerJava/DFS.java
import java.util.*;

class DFS {

    static void run(Cell[][] grid, Cell start, Cell end) {
        run(grid, start, end, 50);
    }

    static void run(Cell[][] grid, Cell start, Cell end, int delay) {

        Stack<Cell> stack = new Stack<>();
        Map<Cell, Cell> parent = new HashMap<>();
        Set<Cell> visited = new HashSet<>();

        stack.push(start);
        visited.add(start);

        while (!stack.isEmpty()) {

            Cell current = stack.pop();

            if (current == end) {
                BFS.reconstructPath(parent, end, start);
                return;
            }

            for (Cell neighbor : BFS.getNeighbors(grid, current)) {

                if (!visited.contains(neighbor) && !neighbor.wall) {

                    stack.push(neighbor);
                    visited.add(neighbor);
                    parent.put(neighbor, current);

                    // do not recolor start/end
                    if (!neighbor.start && !neighbor.end)
                        neighbor.setVisited();

                    try {
                        Thread.sleep(delay);
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                    }
                }
            }
        }
    }
}
